Code restructure

asherbender opened the pull request:
I have mainly cleaned up the code (reduced the number of files, enforced better PEP8 compliance, simplified the API). None  the linear algebra has been modified in any significant way. The interface to the algorithms has changed. This is demonstrated in ./tests.py. My aim was to make it an easier 'blackbox' algorithm to use where as much as possible is handled inside the algorithm.

asherbender committed:
pylint: Cleaned up errors.

    - tests.py
    - __init__.py
    - __util__.py
    - infeng.py
    - tools.py

asherbender committed:
tests.py: Test code is now running.

    - The main problem was the method definition:

          def sim(self, *pred, featfun=None)

      This has been changed to:

          def sim(self, *pred)
              featfun = None

asherbender committed:
tools: tools.py deleted

    - The aim is to boil the 'core' source code down to its bare
      elements. To separate testing code from change detection code,
      tools.py has been eliminated. To facilitate this change the
      following functions have been moved out of tools.py:

          - gendata(), plotprob() and plotbound() moved into tests.py
          - filterdata() and segmentdata() moved into tests.py

asherbender committed:
.gitignore: Git ignore file added to project

asherbender committed:
infeng: suffstat move from __util__.py to infeng.py

    - __util__.py only contains the class suffstat() which is not
      particularly useful outside of the engine() class. To reduce file
      clutter, the __util__.py has been deleted.

asherbender committed:
infeng: accum made a private function

    - accum() is not referenced outside of engine() so it has been
      absorbed into engine() and made a private function.

asherbender committed:
infeng: sufficient statistics validation moved

    - Originally engine() validated the sufficient statistics in
      engine.setparam(). This seemed like an odd place to do it given
      that the sufficient statistics had been decoupled from engine() in
      the class suffstat(). Input validation has been moved to
      suffstat().

    - engine.getparam() and engine.setparam() are no longer required and
      have been removed.

    - engine.init() has been moved into engine.__init__(). The
      instantiation interface has changed to accept the sufficient
      statistics - they are internally passed directly to suffstat()

    - The combination of these changes makes instantiation of the
      inference engine much cleaner (less methods called in the
      process).

    - engine() renamed to the less generic name Bcdm() (Bayesian change
      detection model)

asherbender committed:
infeng: 'Automatic' hypothesis pruning

    - Hypotheses will be pruned at the end of every call to
      bcdm.update() if a maximum number of hypotheses has been set. The
      function can still be called manually.

asherbender committed:
infeng: PREpend segments rather than POSTpend

    - The segmentation method has been modified slightly to grow the
      list from the head, not the tail.

asherbender committed:
infeng: Equation numbers added to some lines of code

    - Equation references from the original 'Bayesian Model-based
      Sequence Segmentation for Inferring Primitives in
      Driving-behavioral Data' have been added

          - These will have to be updated to the new/(most appropriate)
            paper

asherbender committed:
infeng: pred/resp chanded to X and Y

asherbender committed:
infeng: Algorithm changes

    - Stylistic/formatting changes:

        - Changed numpy namespace to np (import numpy as np).
        - Changed 'math' namespace references to 'np' and removed math
          import.
        - changed private variables to 'self.__variable' instead of
          'self.__variable__'.
        - changed hypothesis objects to dictionaries.

    - Interface changes:

        - Specification of rate fun is now done when the algorithm is
          instantiated.
        - No longer need to specify the dimensions of the input/output
          variables. They are either specified by the dimensions of the
          hyperparameters or the first observed data.

    - Extensions:

        - Allow for preserving the order of hypotheses during the trim()
          method. Note that the update() method does not actually
          require that the hypotheses be ordered!

asherbender committed:
infeng: Removed sim() method from algorithm

    - The sim() method is only used in the examples/tests. To reduce
      clutter in the API, it has been removed. The code has been moded
      to the gendata() function in the tests module.

    - Changed 'numpy' import to 'np' in tests.py

asherbender committed:
infeng: segment() recovers hypotheses probabilities

    - The output of segment() now depends on whether the 'maxprod' or
      'sumprod' algorithm is used:

        - If the 'maxprod' algorithm is selected, a list containing the
          segment boundaries is returned.

        - If the 'sumprod' algorithm is selected, a numpy array
          containing the hypothesis probabilities is returned.

    - These changes provide some symmetry to the segmentation
      algorithm. The algorithm is run via calls to update() and
      segment() recovers the final data (whether it is the boundaries or
      the probabilities). The change also moves boiler plate code from
      the users code (such as reconstructing the probabilities) to the
      segmentation class.

asherbender committed:
infeng: Removed filterdata() and segmentdata()

    - Functionality these functions provide have been developed into the
      segmentation class (e.g. recovering the hypothesis probabilities)
      and are now trivial. This code should be implemented by the user.

    - Trivial block_update() function added to allow users to add blocks
      of data. The function makes repeated calls to update() - simply
      saving the user from needing to implement this trivial loop.

asherbender committed:
infeng: Documentation added to algorithm

asherbender committed:
tests: Made plotting code more compact

    - Use of axvspan() and axvline() made code more compact (and allows
      lines/patches to extend to infinity!)

    - Reduced replicated code with a loop.

asherbender committed:
infeng: module renamed to segmentation

asherbender committed:
segmentation: Forgot to add file back

gabrieag commented on .gitignore:
```diff
@@ -1,3 +1,7 @@
+
+# Ignore editor configurations.
+.project
+.pydevproject
```
Could you leave the `*.wpr` and `*.wpu` extensions? I still use the WingWare IDE from time to time.

asherbender replied:
Definitely... 

Sorry. Originally I started editing this thinking we'd only be using local changes in the office.

gabrieag commented on __init__.py:
```diff
@@ -1,5 +1,5 @@
 
 # Import the inference engine.
-from changedetec.infeng import engine
+#from changedetec.infeng import engine
```
Remove dead code.

asherbender replied:
Absolutely!

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+.. codeauthor:: Asher Bender <a.bender@acfr.usyd.edu.au>
+
+"""
+import numpy as np
```
Is this for PEP8 compliance? I prefer the full `numpy` namespace when possible.

asherbender replied:
It is not PEP8 compliance. It is common practice supported by (originating from?) many of the large projects. For instance [documentation in SciPy](http://scipy.org/getting-started.html#an-example-script). The [pylab option](http://ipython.org/ipython-doc/2/api/generated/IPython.core.magics.pylab.html#IPython.core.magics.pylab.PylabMagics.pylab) in ipython imports numpy as np. 

That said, it is not a requirement. If you prefer longer namespacing I can get onboard with that! Although if you do use ipython (I recommend it) **np** is good for the pylab option.

gabrieag replied:
I'm just paranoid about name clashes. But I can't think of a strong argument not to use the `np` binding.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+                np.isfinite(eta) and eta > n - 1.0):
+            msg = 'The shape parameter must be greater than one minus the'
+            msg += ' degrees of freedom.'
+            raise Exception(msg)
```
Are these argument checks really necessary? In the end it's the inference engine that uses these classes. Couldn't we just use `try` + `except` statements when initializing the statistics?

asherbender replied:
Sure! Try-catch is more 'pythonic'. That said, although duck-typing is the python way (and at times a bit dogmatic), I think sanitising user input makes it easier for people to use unfamiliar code. When an exception with a useful error is raised (because the error was anticipated), users do not have to follow a stack trace or debug the code to understand a generic error.

I retained these checks from the original code (although I converted them from asserts to exceptions). Including them incurs a small performance hit as the inputs are checked everytime a hypothesis is created. Although this class could be used outside the inference engine (very unlikely), as you point out the inference engine only uses this class. Originally I was thinking about passing this object into the inference engine at instantiation. Currently there is no point checking the inputs at this level. If they are retained, they should occur once in the inference engine.

gabrieag replied:
How about something like this:

```
try:
    # Initialize the object, allocate space for the
    # sufficient statistics etc.
except:
    # Check all input arguments one by one and 
    # throw an informative error message.
```

asherbender replied:
Unfortunately not!

The try catch approach only works when the code is executed. At this stage they not used (only stored). Any errors in specification will occur later in the code (when the error message is no longer as informative).

The sufficient statisitcs class could be made private. At that point people can use it at their own peril. The code checking can then be moved into the inference engine - it would only be performed once per object (not every time the SS are created).
variable sizes for storage. They are used in operator seldom fails

gabrieag replied:
How do you make a class private? Does it need to be a sub-class of `ChangeDetector`?

asherbender replied:
Privacy in python does not operate like C/C++. There are _conventions_ but these are cultural and not enforced by the interpreter. As you know, in a class you can make a variable "private" by using the double underscore. This causes the python interpreter to [mangle](http://en.wikipedia.org/wiki/Name_mangling#Name_mangling_in_Python) the
variable or method. At a class level, you can use one underscore. In our case the object would become `_MatrixVariateNormalInvGamma`. This is just a flag to other users warning them not to use the object. The auto complete mechanisms used in python IDEs often also tend to ignore these objects and functions.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+class MatrixVariateNormalInvGamma(object):
+    """Matrix-variate normal, inverse-gamma distribution.
+
+    The matrix-variate normal, inverse-gamma distribution is the conjugate
```
`... inverse matrix-variate gamma ...`

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+from scipy import special
+
+
+class MatrixVariateNormalInvGamma(object):
```
+1 for meaningful name.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+
+
+class MatrixVariateNormalInvGamma(object):
+    """Matrix-variate normal, inverse-gamma distribution.
```
Matrix-variate normal, matrix-variate inverse gamma distribution.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+        distribution with observed data, the sufficient statistics refer to the
+        posterior distribution.
+
+        """
```
The sufficient statistics summarize all the information content of the data. Initially, the sufficient statistics contain no information apart from that implied by the prior distribution. As data arrive, the statistics are updated incrementally in order to reflect this new knowledge.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+        sufficient statistics also form the hyper-parameters of the
+        distribution. In a Bayesian context, when the sufficient statistics are
+        referred to as 'hyper-parameters', it implies the sufficient statistics
+        reflect prior information about the distribution. After updating the
```
I think there's a mixup between sufficient statistics and hyper-parameters. The sufficient statistics comprise the block-matrix `S = [(XX, XY), (YX, YY)]`. The hyper-parameters are the parameters of the distribution, i.e. `mu`, `omega`, `sigma` and `eta`. The posterior hyper-parameters can be computed from the sufficient statistics, i.e. `mu, omega, sigma, eta = f(S)`. This `f` is the `param` method.

asherbender replied:
Probably!

I have been looking at the simpler normal, inverse-gamma Bayesian linear model. In that case the goal is to estimate the unknown mean and variance of a Gaussian (over the parameters of a linear model). The hyper-parameters are the sufficient statistics of the NiG prior distribution placed on the Gaussian. Initially they are set to be uninformative. Subsequent observations inform updates of the sufficient statistics of the NiG distribution: S+ = f(S-, X, Y). I need to read up on the matrix variate case - I may have confused things there?

gabrieag replied:
I think the mixup is mainly due to wording. When you say hyper-parameters, you're usually talking about some Bayesian model. This Bayesian model has some prior distribution over its parameters. The _parameters_ of this prior are the _hyper-parameters_ of the Bayesian model.

In this case, the `MatrixGaussianInverseGamma` class doesn't know about the `ChangeDetector` class. Therefore, in this context I find it more appropriate (and less confusing) to talk about `MatrixGaussianInverseGamma`'s _parameters_ (which are `ChangeDetector`'s _hyper-parameters_).

Does this make sense?

asherbender replied:
Ahh yes! Absolutely... I was thinking about it as a system. I can definitely see the ambiguity when you separate them into separate objects.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+            self.__prod[m:, m:] += np.outer(Y, Y)
+            self.__weight += 1
+
+    def logconst(self):
```
Given the standards-complying mood, should we change the name of this method to `getLogConstant`? This way we get to be compatible with PEP8 _and_ with the [Google style guidelines](http://google-styleguide.googlecode.com/svn/trunk/cppguide.html) (at least to some extent).

asherbender replied:
Indeed! Although google has a [style guide for python!](http://google-styleguide.googlecode.com/svn/trunk/pyguide.html). It is quite [PEP8-ish](https://www.python.org/dev/peps/pep-0008/). They both favour CamelCaseClasses, underscore_methods and underscore_variables.

gabrieag replied:
Great! Let's go with that. Unless it clashes with PEP8?

asherbender replied:
Not that I can tell. I am sure there are subtle differences but they certainly aren't obvious. I think Google style guide complements PEP8 and is easier to digest (given the examples).

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+               n * (0.5 * w) * np.log(0.5 * w)
+
+    def param(self):
+        """Return parameters of the posterior distribution."""
```
Return the posterior hyper-parameters.

All the information content of the data is summarized by the sufficient statistics. Hence the hyper-parameters are a function of the sufficient statistics. This is a consequence of the conjugacy of the matrix-variate Gaussian-inverse-Gamma distribution and the linear model.

gabrieag replied:
Scratch hyper-parameters and replace with parameters.

asherbender replied:
Hehe... I am spreading my confusion?

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+               w * np.log(d[m:] / np.sqrt(w)).sum() - \
+               n * (0.5 * w) * np.log(0.5 * w)
+
+    def param(self):
```
`getParameters` or `getHyperParameters`?

gabrieag replied:
`get_parameters`

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+               np.dot(s[m:, m:].transpose(), s[m:, m:]) / w, \
+               w
+
+    def rand(self):
```
`genRandParameters`? Note: we should be clear about the distinction between parameters (i.e. `A` and `B`) and hyper-parameters (i.e. `mu`, `omega', etc.) to avoid confusion.

gabrieag replied:
How about just `simulate`?

asherbender replied:
This is where my knowledge of matrix-variate distributions falls appart somewhat.

I am familiar with the Normal-Wishart as a prior for the multivariate normal distribution. In this case, the parameters (mu, omega, etc) would be the sufficient statistics defining the Normal-Wishart. Given these statistics you can calculate the expected mean and variance of the Gaussian.  If you sample from the Normal-Wishart, you are sampling different (random) multivariate Gaussians. I am not sure how this relates to the matrix-variate version though? It must be similar?

It certainly looks like the returned parameters a and b are the mean and variance of a random multivariate-Gaussian?

gabrieag replied:
Yes, if you sample from the Gauss-Wishart, or a Gauss-inverse-Wishart, you'll get a random mean vector and covariance matrix --the difference between the Wishart and the inverse-Wishart is that they offer different parameterizations of the covariance.

If you sample from the matrix-variate Gauss-inverse-Gamma, you'll get a random [design matrix](http://en.wikipedia.org/wiki/Design_matrix) and covariance matrix. The inverse-Gamma parameterization is similar to the inverse-Wishart (they're both related by a parameter transformation).

So, in short, `A` and `B`, respectively, are random design and covariance matrices.

asherbender replied:
Interesting... Makes me more curious about the matrix variate distributions. I should look at Minka's tech report...

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+                        :py:meth:`.trim`.
+
+    Raises:
+        Exception: If the any of the inputs are an incorrect type.
```
+1 for awesome docstring.

asherbender replied:
We use [Sphinx](http://sphinx-doc.org/) to document our code. It is a bit like Doxygen. With the [Napoleon extension](http://sphinx-doc.org/latest/ext/napoleon.html#module-sphinx.ext.napoleon) you can document code using [Google style](http://sphinx-doc.org/latest/ext/example_google.html#example-google) docstrings.

That said, it has not been set up for this project. I find the Google style docstrings easy to read. It also will allow Sphinx to build documentation later if we decide to set it up.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+                             are initialised during segmentation.
+        eta (float): shape hyper-parameter used when matrix variate, normal
+                     inverse gamma distributions are initialised during
+                     segmentation.
```
Can we somehow group `mu`, `omega`, `sigma` and `eta` to avoid repetition?

asherbender replied:
In the executable code or the documentation? In the executable code we definitely can. 

Not so easily in the documentation. I have just documented each input used to instantiate the object. We could change the input so that it accepts an object which wraps these parameters (like the current MatrixVariateNormalInvGamma object). This would reduce the amount of documentation, in this class, by moving it else where.

gabrieag replied:
What about something a bit more concise, like this:

```
       mu (numpy.array): (M x N) location parameters of the prior distribution.
       omega (numpy.array): (M x M) scale parameters of the prior distribution.
       sigma (numpy.array): (N x N) dispersion parameters of the prior distribution.
       eta (float): shape parameter of the prior distribution.
```

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+                         1. Segments are MORE likely to be created with values
+                         closer to zero. Segments are LESS likely to form with
+                         values closer to 1. Alternatively, ratefun can be set
+                         to an executable hazard function.
```
The hazard function must accept non-negative integers and return non-negative floating-point numbers.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+                        limit the number of hypotheses. If set to ``None``, no
+                        pruning will NOT take place after updates, however,
+                        pruning can be initiated manually by calling
+                        :py:meth:`.trim`.
```
featfun (callable): Feature functions for basis function expansion. Feature functions provide additional flexibility by mapping the predictor variables to an intermmediate feature space, thus allowing the user to model non-linear relationships.

Maybe we should change `featfun` to `basisfun`.

asherbender replied:
Definitely! I am more familiar with the basis function / design matrix nomenclature

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+        self.__ratefun = ratefun if callable(ratefun) else lambda x: ratefun
+
+    def __initialise_algorithm(self, m, n):
+        """Initialise the Bcdm algorithm."""
```
What's Bcmd?

asherbender replied:
Yeah.... I wasnt sure what to do here. I wanted something more descriptive than _engine_. Your other algorithm was called BMOSS but I didn't know if they should be kept distinct or given the same name. I opted for keeping them distinct and turned the more generic _Bayesian change detection model_ into an acronym.

It by no means was the best choice! I am definitely open to better naming.

gabrieag replied:
`LinearChangeDetector`? It's kind of bland, but that's all I can think of at 8:48 pm.

asherbender replied:
Hmmm... Some words:
- Linear model
- Change point detection
- Piecewise regression

gabrieag replied:
We'll have to think about this.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+        if self.__m is None:
+            self.__m = m
+        elif self.__m != m:
+            msg = 'Expected %i dimensions in the predictor variable.' % m
```
Shouldn't we be using `str.format`?

asherbender replied:
Probably! Yes. Sorry. Again, I think I did this early on before I had a view to contribute. I simply liked the older notation better since it can be more compact. But 'format' is more flexible.

gabrieag replied:
Also for Python 3 compatibility.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+                              'distribution': stat,
+                              'log_constant': stat.logconst()}]
+
+    def __accum(self, x, y):
```
Make static.

asherbender replied:
Agreed. I also tried to think of a better name but couldnt really...

gabrieag replied:
How about `__softmax`? This is really just the binary version of the [softmax function](http://en.wikipedia.org/wiki/Softmax_function).

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+        # Iterate through the rows of the input-output data updating the
+        # hypotheses in the model.
+        for i in range(X.shape[0]):
+            self.update(X[i, :], Y[i, :])
```
I don't think this method is necessary. Whether there's a single pair of input-output data or multiple pairs, this should be transparent to the algorithm. The single/multiple cases get handled downstream by the sufficient statistics.

asherbender replied:
I thought the algorithm had to iterate through each point to determine if there was a break in the middle of a block? It sounds like this is not the case? For instance can you perform a single update using the entire dataset? If so this method is definitely worth deleting! (I tried this early on but it crashed - perhap in the plotting?). 

I definitely agree the better course is to have an update method where the data size is transparent to the algorithm.

gabrieag replied:
The granularity of the segmentation (i.e. the number of input-output observations in a single block) is entirely up to the user. If you call `update` with blocks of _n_ observations each, the segmentation would have a minimum resolution of _n_ time steps. This is handy for high-frequency observations like IMU data (which is what I used for the experiments in the paper). Of course, there's no way to know whether there's a change point inside a block.

gabrieag replied:
PS: If you fed the entire data all at once, things should still work. If they don't, there must be something wrong.

asherbender replied:
Hmm I assumed that how the blocks worked... Which is still why I am unsure about how feeding in an entire dataset works. If you feed the data in all at once (one update) wont the segmentation have a minimum resolution equal to the length of the data set?

gabrieag replied:
It should work in the sense that it shouldn't crash. It will give meaningless results (since the dataset will be considered as a single big chunk).

asherbender replied:
Ahh... yes. Those are two very different things. Correct, I don't believe it crashed but the results were indeed meaningless.  This is why I wrote `block_update()` originally. It was just to stop users from having to write an external loop. Loops are fairly compact in python so removing the method is not a big inconvenience to users (it could be added as an option to `update()`?). Though we should make a note in the documentation that the resolution of the segmentation is linked to how many data points are inserted during calls to `update()`.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+            self.update(X[i, :], Y[i, :])
+
+    def update(self, X, Y):
+        """Update model with a single observation.
```
Again, I don't think we need to make any distinction.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+                              'count': 0,
+                              'log_probability': 0.0,
+                              'distribution': stat,
+                              'log_constant': stat.logconst()}]
```
Nice. It never occured to me to store the hypotheses as a `dict`.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+        # Update hypotheses by updating each maxtrix variate, normal inverse
+        # gamma distribution over the linear models.
+        self.__update_count += 1
+        for hypotheses in self.__hypotheses:
```
`for hypothesis in self.__hypotheses:` we're iterating over hypothes**e**s, one hypothes**i**s at a time.

asherbender replied:
Ha! Yup... mindless.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+
+            # Update and accumulate the log-probabilities.
+            hypotheses['log_probability'] += np.log1p(-hazard) + log_density
+            logsum = self.__accum(logsum, hypotheses['log_probability'])
```
I didn't go over this in detail. I assume the equations haven't changed with respect to the original implementation.

asherbender replied:
My aim was not to change any core code. I can double check this...

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+        # hypotheses.
+        if self.__alg__ == 'maxprod':
+            loglik = logmax
+            self.__index.append(ind)
```
How about renaming `__index` to something more meaningful like `__max_indices`?

asherbender replied:
Definitely more descriptive

gabrieag replied:
On second thought, how about we call it `__counts`? That way, when backtracking we can say `count = self.__counts[index - 1]`.

asherbender replied:
sure!

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+                iteration.append((hypothesis['count'],
+                                  hypothesis['log_probability']))
+
+            self.__probabilities.append(iteration)
```
Why do we need `__probabilities`? Isn't all the information already stored in `__hypotheses`?

asherbender replied:
There is probably a better way to do this.

Originally the probabilities were updated in **filterdata** of tools.py. That was done using:

```
for i in range(k):

    # Update the segmentation hypotheses given the data, one point at a
    # time.
    eng.update(pred[i, :], resp[i, :], **arg)

    # Limit the number of hypotheses.
    eng.trim()

    # Update the probabilities.
    for j, alpha in eng.state():
        prob[j, i + 1] = alpha
```

In an attempt to internalise that code into the inference engine class, I added another variable. The aim was to prevent users from having to duplicate the _# Update the probabilities_ block. Although this repetition would not be necessary if they ignored the inference class and just used filterdata...

gabrieag replied:
So you're storing the probabilities of each of the hypotheses at each time step? I'm just worried about memory limitations when you want to run this online, indefinitely. I think it's best to have a convenience function (like `filterdata`) for this.

asherbender replied:
hmm... this is true. An alternative might be to allow an optional limitation on the amount of data stored? A rotating buffer of N-observations where old probabilities or segments are pushed off the head? This would push the memory management back to the user but they wouldn't have to execute code externally.

gabrieag replied:
What about a wrapper class? This way we don't bloat the code and avoid burdening the inference engine unnecessarily.

asherbender replied:
Definitely a viable option!

I guess it would also keep the objects a bit more atomic... If you follow that pattern, you could probably even implement the majority of the inference engine in one class and inherit from that to implement the max-product and max-sum algorithms. Doing so would reduce the need for conditionals to switch behaviours.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+
+            self.__probabilities.append(iteration)
+
+    def trim(self, minprob=1.0e-6, maxhypot=20):
```
`trimHypotheses`?

asherbender replied:
trim_hypotheses?

gabrieag replied:
Yes.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+        for hypot in self.__hypotheses:
+            hypot['log_probability'] -= logsum
+
+    def segment(self):
```
How about `infer`? This should suit both purposes.

asherbender replied:
Sure!

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+        array. Rows in the array represent hypotheses and columns in the array
+        represent data points in the time-series.
+
+        Returns:
```
This method returns the inference results In the case of the max-product algorithm, the method returns the most likely segmentation ...

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+                segment_boundaries.insert(0, index)
+                count = self.__index[index - 1]
+
+            return segment_boundaries
```
Again, I haven't checked the details here. I assume it's the same as in the original code.

asherbender replied:
Actually I modified this a little. I changed the code to add segment boundaries to the head of the list (rather than the tail) so that segment_boundaries.reverse() would not need to be called. Other than that I believe it is the same.

gabrieag commented on tests.py:
```diff
@@ -1,107 +1,192 @@
+#!/usr/bin/python
```
I'll look at this another day. It would be good to have some sort of unit test, e.g. some data with known segmentation results. I'll think about this.

asherbender replied:
Unit testing is always a good thing. They are also easy to set up - I am happy to help out. Python even offers '[doctests](https://docs.python.org/2/library/doctest.html)' which can be integrated with Sphix.

gabrieag replied:
What if we kept a dataset, generated with the original code, and had a `nose` unit test that runs the current implementation and compares the results to the known results.

gabrieag commented on segmentation.py:
```diff
@@ -0,0 +1,575 @@
+        """
+
+        # Skip pruning if less hypotheses exist than the maximum allowed.
+        if len(self.__hypotheses) <= maxhypot:
```
I guess we should also review variable naming at some point (`max_hypotheses` instead of `maxhypot`). But this'll be in a future pull request.

gabrieag commented on tests.py:
```diff
@@ -1,107 +1,192 @@
+
+    # Set uninformative prior for the shape parameter.
+    if eta is None:
+        eta = n
```
These are not really non-informative priors. They're ambiguous, but somewhat informative. The only way to have a non-informative prior would be to set `omega = epsilon * numpy.eye(m)` and `eta = n + epsilon - 1`, with a really small `epsilon`.

gabrieag commented on tests.py:
```diff
@@ -1,107 +1,192 @@
+    bound = random.permutation(np.arange(k - 1) + 1)
+    bound = np.concatenate([np.array([0]),
+                            np.sort(bound[:l-1]),
+                            np.array([k])])
```
There are other ways to do this, e.g. if you want to generate segments of approximately the same. Let me know if you need something like this.

asherbender replied:
I can't imagine needing it for now? It suits the purpose for generating random data.

gabrieag commented on tests.py:
```diff
@@ -1,107 +1,192 @@
+    """Plot segment boundaries as vertical lines."""
+
+    for i in range(len(segments)):
+        plt.axvline(x[segments[i]], **args)
```
Everything is nicely organized into functions. Awesome!

asherbender replied:
The matplotlib functions also plot to infinity which makes panning around the figure nicer.

gabrieag commented on tests.py:
```diff
@@ -1,107 +1,192 @@
-    loweraxes.set_ylabel('Hypothesis probability')
+    # Set parameters for generating the data.
+    gainparam = 0.5
+    noiseparam = 5.0
```
Going to have dinner now. I'll continue from here later.

asherbender replied:
Thanks for all the input!

asherbender committed:
gitignore: WingWare IDE temporary files added

asherbender committed:
init: Removed commented out / dead code

asherbender committed:
segmentation: Updated short doc string of MatrixVariateNormalInvGamma

asherbender committed:
segmentation: Clarified sufficient statistics documentaiton

    - Updated the documentation in the
      MatrixVariateNormalInvGamma.update() method. The process of
      updating the sufficient statistics in the has been clarified.  The
      class is a distinct object representing a distribution. It does
      not represent a Bayesian model. References to hyper-parameters
      raise this ambiguity. The confusing references to
      hyper-parameters have been removed.

    - Updated documentation in the MatrixVariateNormalInvGamma.param()
      method. Same reason.

asherbender committed:
segmentation: PEP8ification logconst->log_const

    - MatrixVariateNormalInvGamma.logconst() changed to
      MatrixVariateNormalInvGamma.log_const

asherbender committed:
segmentation: Descriptification params->parameters

    - MatrixVariateNormalInvGamma.params() changed to
      MatrixVariateNormalInvGamma.parameters()

asherbender committed:
segmentation: Updated doc string of Bcdm

    - Made documentation of prior distribution parameters more compact.
    - Update documentation of the hazard function (inputs and outputs).
    - Added documentation for 'featfun'

asherbender committed:
segmentation: Renamed featfun->basisfun

asherbender committed:
segmentation: Upgrade to str.format()

    - In formated errors of Bcdm()

asherbender committed:
segmentation: rename __accum->__soft_max

    - Bcdm.__accum() changed to Bcdm.__soft_max()

asherbender committed:
segmentation: Renamed methods/variables variables

    - hypotheses -> hypothesis In main loop of Bcdm.update()
    - Bcdm.__index -> Bcdm.__counts
    - Bcdm.segment() -> Bcdm.infer()

asherbender committed:
test: Changed comments.

    - Reflect that the prior parameters used in random_segments are not
      uninformative priors. They are simply the default priors.

asherbender committed:
segmentation: Bug fix

    - Introduced in: 616215ea28ef051211b9d61ab304aadb8252f8d2

    - In the max-product algorithm, the log-likelihood is updated to the
      maximum log-likelihood at the end of an update
      forward-pass. Previously this was done too late and new hypotheses
      were being created with incorrect log-likelihood data. The
      log-likelihood is now updated before new hypotheses are created.

asherbender committed:
segmentation: Set initial hypothesis probability

    - Set probability of the first hypothesis to one not zero.

asherbender committed:
segmentation: Changed default hypothesis probabilities

    - Set default posterior segmentation hypothesis to one not zero.

asherbender committed:
tests: Added non-sinusoid waves to test routines

    - The non-sinusoid waves have known characteristics, making it a
      good case for debugging.

    - Reverted changes to probability initialisation

asherbender committed:
segmentation: Re-instated time-variant basis functions

    - Re-instated time-variant basis functions. Each hypothesis now
      stores the default basis function or the one supplied during a
      call to update.

    - Renamed self.__basisfun to self.__basisfunc

    - Created function for initialising new hypotheses.

    - Provided a weaker prior on the Gaussian dispersion parameter in
      the non-sinusoidal test. This allows the algorithm to better
      detect changes in the triangle-wave form.

asherbender committed:
segmentation: Removed block update method

    - User now have to iterate through the data themselves.
    - Non-sinusoidal test uses time-variant basis functions.

asherbender committed:
segmentation: Bug fix, basis function expansion

    - The algorithm was ignoring the size of the design matrix (input after
      basis function expansion) when intialising the priors. This has
      been corrected.

    - The algorithm is now rather finicky/fragile about how data is
      input. Pay careful attention to how the dimensionality of the data
      is formatted before feeding it to the update() method.

gabrieag commented:
@asherbender good to go?

gabrieag commented:
@asherbender I just gave you push access. Feel free to merge (you should do the honours)!

asherbender committed:
MatrixVariateNormalInvGamma: .transpose() -> .T

    - Change of notation to keep the equations terse
